sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+  bn.addm  w22, w20, w31
+
+  ret
+
+/**
+ * Multiply two field elements and reduce modulo p.
+ *
+ * Returns c = (a * b) mod p.
+ *
+ * The inputs a and b must be at most 2^255 bits, although it is not necessary
+ * for them to be fully reduced modulo p. This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  w22: a, first operand, a < 2^255
+ * @param[in]  w23: b, second operand, b < 2^255
+ * @param[in]  MOD: p, modulus = 2^255 - 19
+ * @param[in]  w31: all-zero
+ * @param[out] w22: c, result
+ *
+ * clobbered registers: w18, w20 to w22
+ * clobbered flag groups: FG0
+ */
+.globl fe_mul
+fe_mul:
+  /* Compute the raw 510-bit product.
+     [w21:w20] <= a * b */
+  bn.mulqacc.z          w22.0, w23.0,  0
+  bn.mulqacc            w22.1, w23.0, 64
+  bn.mulqacc.so  w20.L, w22.0, w23.1, 64
+  bn.mulqacc            w22.2, w23.0,  0
+  bn.mulqacc            w22.1, w23.1,  0
+  bn.mulqacc            w22.0, w23.2,  0
+  bn.mulqacc            w22.3, w23.0, 64
+  bn.mulqacc            w22.2, w23.1, 64
+  bn.mulqacc            w22.1, w23.2, 64
+  bn.mulqacc.so  w20.U, w22.0, w23.3, 64
+  bn.mulqacc            w22.3, w23.1,  0
+  bn.mulqacc            w22.2, w23.2,  0
+  bn.mulqacc            w22.1, w23.3,  0
+  bn.mulqacc            w22.3, w23.2, 64
+  bn.mulqacc.so  w21.L, w22.2, w23.3, 64
+  bn.mulqacc.so  w21.U, w22.3, w23.3,  0
+
+  /* Reduce modulo p.
+     w22 <= [w21:w20] mod p = (a * b) mod p */
+  jal    x1, fe_reduce
+
+  ret
+
+/**
+ * Square a field element and reduce modulo p.
+ *
+ * Returns c = (a * a) mod p.
+ *
+ * The input a must be at most 2^255 bits, although it is not necessary for it
+ * to be fully reduced modulo p. This specialized squaring procedure is
+ * slightly faster than fe_mul, because duplicated partial products can be
+ * multiplied by two instead of being computed separately.  By optimizing for
+ * this special case, we can use 10 multiplications and 4 additions instead of
+ * 16 multiplications to compute the raw product.
+ *
+ * Note that this is only 2 instructions faster than fe_mul; if we need to cut
+ * down on code size, we could try not using a specialized square. However,
+ * this routine is called many, many times (especially in the inverse
+ * computation) so those two instructions might add up to quite a bit in the
+ * end.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  w22: a, operand, a < 2^255
+ * @param[in]  MOD: p, modulus = 2^255 - 19
+ * @param[in]  w31: all-zero
+ * @param[out] w22: c, result
+ *
+ * clobbered registers: w17, w18, w20 to w22
+ * clobbered flag groups: FG0
+ */
+.globl fe_square
+fe_square:
+  /* Compute the partial products that do NOT need to be multiplied by 2.
+     [w21:w20] <= a0^2 + (a1^2 << 128) + (a2^2 << 256) + (a3^2 << 384) */
```

Comment:
Ah you use the same notation below for the partial products. There it would get a bit clunky with the bit ranges. You could define a0 to a3 in the beginning (although it is pretty obvious)

Reply:
Good point, I'll define the `a0-a3` notation in the beginning here.